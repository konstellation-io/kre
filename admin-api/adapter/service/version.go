package service

import (
	"context"
	"fmt"
	"time"

	"google.golang.org/grpc"

	"github.com/konstellation-io/kre/admin-api/adapter/config"
	"github.com/konstellation-io/kre/admin-api/adapter/service/proto/versionpb"
	"github.com/konstellation-io/kre/admin-api/domain/entity"
	"github.com/konstellation-io/kre/admin-api/domain/usecase/logging"
)

type K8sVersionClient struct {
	cfg    *config.Config
	client versionpb.VersionServiceClient
	logger logging.Logger
}

func NewK8sVersionClient(cfg *config.Config, logger logging.Logger) (*K8sVersionClient, error) {
	cc, err := grpc.Dial(cfg.Services.K8sManager, grpc.WithInsecure())
	client := versionpb.NewVersionServiceClient(cc)
	if err != nil {
		return nil, err
	}

	return &K8sVersionClient{
		cfg,
		client,
		logger,
	}, nil
}

// StartVersion creates resources in k8s
func (k *K8sVersionClient) Start(runtime *entity.Runtime, version *entity.Version) error {
	wf := make([]*versionpb.Version_Workflow, len(version.Workflows))

	for i, w := range version.Workflows {
		nodes := make([]*versionpb.Version_Workflow_Node, len(w.Nodes))
		for j, n := range w.Nodes {
			nodes[j] = &versionpb.Version_Workflow_Node{
				Id:    n.ID,
				Name:  n.Name,
				Image: n.Image,
				Src:   n.Src,
			}
		}
		edges := make([]*versionpb.Version_Workflow_Edge, len(w.Edges))
		for k, e := range w.Edges {
			edges[k] = &versionpb.Version_Workflow_Edge{
				Id:       e.ID,
				FromNode: e.FromNode,
				ToNode:   e.ToNode,
			}
		}

		wf[i] = &versionpb.Version_Workflow{
			Id:         w.ID,
			Name:       w.Name,
			Entrypoint: w.Entrypoint,
			Nodes:      nodes,
			Edges:      edges,
		}
	}

	configVars := make([]*versionpb.Version_Config, len(version.Config.Vars))
	for i, c := range version.Config.Vars {
		configVars[i] = &versionpb.Version_Config{
			Key:   c.Key,
			Value: c.Value,
		}
	}

	req := versionpb.Request{
		Version: &versionpb.Version{
			Id:        version.ID,
			Name:      version.Name,
			Namespace: runtime.GetNamespace(),
			Config:    configVars,
			Entrypoint: &versionpb.Version_Entrypoint{
				ProtoFile: version.Entrypoint.ProtoFile,
				Image:     version.Entrypoint.Image,
				Src:       version.Entrypoint.Src,
			},
			Workflows: wf,
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	_, err := k.client.Start(ctx, &req)
	if err != nil {
		return err
	}

	return nil
}

func (k *K8sVersionClient) Stop(runtime *entity.Runtime, version *entity.Version) error {
	req := versionpb.Request{
		Version: &versionpb.Version{
			Name:      version.Name,
			Namespace: runtime.GetNamespace(),
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	_, err := k.client.Stop(ctx, &req)
	if err != nil {
		return fmt.Errorf("stop version '%s' error: %w", version.Name, err)
	}

	return nil
}

func (k *K8sVersionClient) UpdateConfig(runtime *entity.Runtime, version *entity.Version) error {
	configVars := make([]*versionpb.Version_Config, len(version.Config.Vars))
	for x, c := range version.Config.Vars {
		configVars[x] = &versionpb.Version_Config{
			Key:   c.Key,
			Value: c.Value,
		}
	}

	req := versionpb.Request{
		Version: &versionpb.Version{
			Name:      version.Name,
			Namespace: runtime.GetNamespace(),
			Config:    configVars,
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Second)
	defer cancel()

	_, err := k.client.UpdateConfig(ctx, &req)
	if err != nil {
		return err
	}

	return nil
}

func (k *K8sVersionClient) Unpublish(runtime *entity.Runtime, version *entity.Version) error {
	req := versionpb.Request{
		Version: &versionpb.Version{
			Name:      version.Name,
			Namespace: runtime.GetNamespace(),
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	_, err := k.client.Unpublish(ctx, &req)
	if err != nil {
		return err
	}

	return nil
}

func (k *K8sVersionClient) Publish(runtime *entity.Runtime, version *entity.Version) error {

	req := versionpb.Request{
		Version: &versionpb.Version{
			Name:      version.Name,
			Namespace: runtime.GetNamespace(),
		},
	}

	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
	defer cancel()

	_, err := k.client.Publish(ctx, &req)
	if err != nil {
		return err
	}

	return nil
}
