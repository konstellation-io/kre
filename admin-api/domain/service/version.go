package service

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/service_${GOFILE} -package=mocks

import "github.com/konstellation-io/kre/admin-api/domain/entity"

type VersionService interface {
	Start(runtime *entity.Runtime, version *entity.Version) error
	Stop(runtime *entity.Runtime, version *entity.Version) error
	Unpublish(runtime *entity.Runtime, version *entity.Version) error
	Publish(runtime *entity.Runtime, version *entity.Version) error
	UpdateConfig(runtime *entity.Runtime, version *entity.Version) error
}
