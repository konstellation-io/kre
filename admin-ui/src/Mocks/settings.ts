import { loader } from 'graphql.macro';

const GetUserActivityQuery = loader(
  '../Graphql/queries/getUserActivity.graphql'
);
const GetDomainsQuery = loader('../Graphql/queries/getDomains.graphql');
const GetExpTimeQuery = loader('../Graphql/queries/getExpirationTime.graphql');
const UpdateSessionLifetimeMutation = loader(
  '../Graphql/mutations/updateVersionConfiguration.graphql'
);
const UpdateDomainsMutation = loader(
  '../Graphql/mutations/updateDomains.graphql'
);

export const domainMock = {
  request: {
    query: GetDomainsQuery
  },
  result: {
    data: {
      settings: {
        authAllowedDomains: ['domain.1', 'domain.2', 'domain.3.sample']
      }
    }
  }
};

export const usersActivityMock = {
  request: {
    query: GetUserActivityQuery
  },
  result: {
    data: {
      userActivityList: [
        {
          user: { email: 'user1@domain.com' },
          type: 'LOGIN',
          vars: [],
          date: '2019-11-28T15:28:01+00:00'
        },
        {
          user: { email: 'user2@domain.com' },
          type: 'LOGIN',
          vars: [],
          date: '2019-11-27T15:28:01+00:00'
        },
        {
          user: { email: 'user3@domain.com' },
          type: 'LOGIN',
          vars: [],
          date: '2019-11-26T15:28:01+00:00'
        }
      ]
    }
  }
};

export const expirationTimeMock = {
  request: {
    query: GetExpTimeQuery
  },
  result: {
    data: {
      settings: {
        sessionLifetimeInDays: 45
      }
    }
  }
};

export const addAllowedDomainMock = {
  request: {
    query: UpdateDomainsMutation,
    variables: {
      input: {
        authAllowedDomains: [
          'domain.1',
          'domain.2',
          'domain.3.sample',
          'intelygenz.com'
        ]
      }
    }
  },
  result: {
    data: {
      updateSettings: {
        authAllowedDomains: [
          'domain.1',
          'domain.2',
          'domain.3.sample',
          'intelygenz.com'
        ]
      }
    }
  }
};

export const updateExpirationTime = {
  request: {
    query: UpdateSessionLifetimeMutation,
    variables: { input: { sessionLifetimeInDays: 10 } }
  },
  result: {
    data: {
      setSettings: {
        settings: {
          sessionLifetimeInDays: 12
        }
      }
    }
  }
};
