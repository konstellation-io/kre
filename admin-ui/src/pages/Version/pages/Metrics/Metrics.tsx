import React, { useEffect, useState, useCallback } from 'react';
import DashboardHeader from './components/DashboardHeader/DashboardHeader';
import Charts from './components/Charts/Charts';
import Message from '../../../../components/Message/Message';
import { loader } from 'graphql.macro';
import styles from './Metrics.module.scss';
import {
  GetVersionConfStatus_versions,
  GetVersionConfStatus_runtime
} from '../../../../graphql/queries/types/GetVersionConfStatus';
import { useQuery } from '@apollo/react-hooks';
import {
  GetMetrics,
  GetMetricsVariables
} from '../../../../graphql/queries/types/GetMetrics';
import SpinnerCircular from '../../../../components/LoadingComponents/SpinnerCircular/SpinnerCircular';
import ErrorMessage from '../../../../components/ErrorMessage/ErrorMessage';
import { useParams } from 'react-router-dom';
import { VersionRouteParams } from '../../../../constants/routes';
import { useForm } from 'react-hook-form';
import moment, { Moment } from 'moment';

const GetMetricsQuery = loader(
  '../../../../graphql/queries/getMetrics.graphql'
);

type FormData = {
  startDate: Moment;
  endDate: Moment;
};

const DEFAULT_DATES: FormData = {
  startDate: moment()
    .subtract(30, 'days')
    .startOf('day'),
  endDate: moment().endOf('day')
};

type Props = {
  runtime?: GetVersionConfStatus_runtime;
  version?: GetVersionConfStatus_versions;
};

function Metrics({ runtime, version }: Props) {
  const [viewAllData, setViewAllData] = useState<boolean>(false);

  const { versionId, runtimeId } = useParams<VersionRouteParams>();
  const { data, loading, error, refetch } = useQuery<
    GetMetrics,
    GetMetricsVariables
  >(GetMetricsQuery, {
    variables: {
      versionId,
      runtimeId,
      startDate: DEFAULT_DATES.startDate.toISOString(),
      endDate: DEFAULT_DATES.endDate.toISOString()
    }
  });

  const { register, watch, setValue, handleSubmit } = useForm<FormData>({
    defaultValues: {
      startDate: DEFAULT_DATES.startDate,
      endDate: DEFAULT_DATES.endDate
    }
  });

  const [expanded, setExpanded] = useState<string>('');
  function toggleExpanded(nodeId: string): void {
    setExpanded(expanded ? '' : nodeId);
  }

  useEffect(() => {
    register({ name: 'startDate' });
    register({ name: 'endDate' });
  }, [register]);

  const submit = useCallback(() => {
    handleSubmit(({ startDate, endDate }: FormData) => {
      refetch({
        runtimeId,
        versionId,
        startDate: startDate.toISOString(),
        endDate: endDate.toISOString()
      });
    })();
  }, [handleSubmit, refetch, runtimeId, versionId]);

  // Submits the form every time 'endDate' is updated
  const endDate = watch('endDate');
  useEffect(() => {
    if (endDate) submit();
  }, [endDate, submit]);

  function getContent() {
    if (loading) return <SpinnerCircular />;
    if (error || !data) return <ErrorMessage />;
    if (data && data.metrics === null)
      return <Message text="There is no data for the selected dates" />;

    return (
      <Charts
        data={data}
        expanded={expanded}
        toggleExpanded={toggleExpanded}
        viewAllData={viewAllData}
      />
    );
  }

  return (
    <div className={styles.container}>
      <DashboardHeader
        runtimeName={runtime && runtime.name}
        versionName={version && version.name}
        value={watch}
        onChange={setValue}
        submit={submit}
        viewAllData={viewAllData}
        setViewAllData={setViewAllData}
      />
      {getContent()}
    </div>
  );
}

export default Metrics;
