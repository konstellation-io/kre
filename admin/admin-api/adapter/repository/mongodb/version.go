package mongodb

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"

	"github.com/konstellation-io/kre/admin/admin-api/adapter/config"
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase/logging"
)

type VersionRepoMongoDB struct {
	cfg        *config.Config
	logger     logging.Logger
	collection *mongo.Collection
}

func NewVersionRepoMongoDB(
	cfg *config.Config,
	logger logging.Logger,
	client *mongo.Client,
) *VersionRepoMongoDB {
	collection := client.Database(cfg.MongoDB.DBName).Collection("versions")
	return &VersionRepoMongoDB{
		cfg,
		logger,
		collection,
	}
}

func (r *VersionRepoMongoDB) Create(userID string, newVersion *entity.Version) (*entity.Version, error) {
	newVersion.ID = primitive.NewObjectID().Hex()
	newVersion.CreationDate = time.Now().UTC()
	newVersion.CreationAuthor = userID
	newVersion.Status = entity.VersionStatusStopped
	res, err := r.collection.InsertOne(context.Background(), newVersion)
	if err != nil {
		return nil, err
	}

	newVersion.ID = res.InsertedID.(string)
	return newVersion, nil
}

func (r *VersionRepoMongoDB) GetByID(id string) (*entity.Version, error) {
	v := &entity.Version{}
	filter := bson.D{{"_id", id}}

	err := r.collection.FindOne(context.Background(), filter).Decode(v)
	if err == mongo.ErrNoDocuments {
		return v, usecase.ErrVersionNotFound
	}

	return v, err
}

func (r *VersionRepoMongoDB) GetByIDs(ids []string) ([]*entity.Version, []error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"_id": bson.M{"$in": ids}})
	if err != nil {
		return versions, []error{err}
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, []error{err}
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) Update(version *entity.Version) error {
	_, err := r.collection.ReplaceOne(context.Background(), bson.M{"_id": version.ID}, version)
	if err != nil {
		return err
	}

	return nil
}

func (r *VersionRepoMongoDB) GetByRuntime(runtimeID string) ([]*entity.Version, error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"runtimeId": runtimeID})
	if err != nil {
		return versions, err
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, err
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) SetHasDoc(ctx context.Context, versionID string, hasDoc bool) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"hasDoc": hasDoc}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}

func (r *VersionRepoMongoDB) SetStatus(ctx context.Context, versionID string, status entity.VersionStatus) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"status": status}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}
