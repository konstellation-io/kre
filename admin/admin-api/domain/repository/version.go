package repository

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/repo_${GOFILE} -package=mocks

import (
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
)

type VersionRepo interface {
	Create(userID string, version *entity.Version) (*entity.Version, error)
	GetByID(id string) (*entity.Version, error)
	GetByIDs(ids []string) ([]*entity.Version, []error)
	Update(version *entity.Version) error
	GetByRuntime(runtimeID string) ([]*entity.Version, error)
}
