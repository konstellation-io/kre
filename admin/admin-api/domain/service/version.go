package service

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/service_${GOFILE} -package=mocks

import (
	"context"

	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
)

type VersionService interface {
	Start(context.Context, *entity.Runtime, *entity.Version) error
	Stop(context.Context, *entity.Runtime, *entity.Version) error
	Unpublish(*entity.Runtime, *entity.Version) error
	Publish(*entity.Runtime, *entity.Version) error
	UpdateConfig(*entity.Runtime, *entity.Version) error
}
