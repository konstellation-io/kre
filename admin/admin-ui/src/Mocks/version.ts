import { RuntimeStatus, VersionStatus } from 'Graphql/types/globalTypes';

import { loader } from 'graphql.macro';

const GetConfigurationVariablesQuery = loader(
  'Graphql/queries/getConfigurationVariables.graphql'
);
const GetVersionWorkflowsQuery = loader(
  'Graphql/queries/getVersionWorkflows.graphql'
);
const GetVersionNodeStatusQuery = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const NodeStatusSubscription = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const PublishVersionMutation = loader(
  'Graphql/mutations/publishVersion.graphql'
);
const StartVersionMutation = loader('Graphql/mutations/startVersion.graphql');
const StopVersionMutation = loader('Graphql/mutations/stopVersion.graphql');
const UnpublishVersionMutation = loader(
  'Graphql/mutations/unpublishVersion.graphql'
);

export const runtime = {
  id: 'runtimeId',
  name: 'runtimeName',
  description: 'Some Description',
  creationAuthor: { email: 'some@user.com' },
  status: RuntimeStatus.STARTED,
  __typename: 'Runtime'
};
export const version = {
  __typename: 'Version',
  id: 'versionId',
  name: 'versionName',
  description: 'versionDescription',
  status: VersionStatus.STARTED,
  creationDate: '2020/01/01',
  creationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  publicationDate: '2020/01/01',
  publicationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  config: {
    vars: [],
    completed: false
  }
};
export const confVarsMock = {
  request: {
    query: GetConfigurationVariablesQuery
  },
  result: {
    data: {
      version: {
        status: 'STARTED',
        config: {
          completed: true,
          vars: [
            {
              key: 'var2',
              value: 'value1',
              type: 'VARIABLE'
            },
            {
              key: 'var1',
              value: 'value2',
              type: 'VARIABLE'
            },
            {
              key: 'var3',
              value: 'value3',
              type: 'FILE'
            }
          ]
        }
      }
    }
  }
};

export const workflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  result: {
    data: {
      version: {
        name: 'version01',
        status: 'STARTED',
        configurationCompleted: true,
        workflows: [
          {
            name: 'workflow01',
            nodes: [
              {
                id: 'node01',
                name: 'one node',
                status: 'STARTED'
              },
              {
                id: 'node02',
                name: 'another node',
                status: 'STARTED'
              }
            ],
            edges: [
              {
                id: 'edge01',
                fromNode: 'node01',
                toNode: 'node02'
              }
            ]
          }
        ]
      }
    }
  }
};

export const errorMorkflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  error: 'Some error'
};

export const nodeStatus = {
  request: {
    query: GetVersionNodeStatusQuery
  },
  result: {
    data: {
      versionNodeStatus: [
        {
          date: '2020-01-01',
          nodeId: 'nodeId1',
          status: 'STARTED',
          message: 'some message 1'
        },
        {
          date: '2020-01-01',
          nodeId: 'nodeId2',
          status: 'STARTED',
          message: 'some message 2'
        }
      ]
    }
  }
};

export const nodeStatusMock = {
  request: {
    query: NodeStatusSubscription
  },
  result: {
    data: {
      versionNodeStatus: {
        date: '2020/01/01',
        nodeId: 'node02',
        status: 'STARTED',
        message: 'some message'
      }
    }
  }
};

export const publishVersionMock = {
  request: {
    query: PublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const startVersionMock = {
  request: {
    query: StartVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const stopVersionMock = {
  request: {
    query: StopVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const unpublishVersionMock = {
  request: {
    query: UnpublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};
