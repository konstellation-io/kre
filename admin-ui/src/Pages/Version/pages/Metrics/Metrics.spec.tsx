import { runtime, version } from 'Mocks/version';

import Metrics from './Metrics';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('react-router', () => ({
  useParams: jest.fn(() => ({ runtimeId: '', versionId: '' }))
}));
jest.mock('@apollo/react-hooks', () => ({
  useQuery: jest.fn(() => ({
    data: {},
    loading: false,
    error: '',
    refetch: jest.fn()
  }))
}));

// TODO: test GraphQL data
describe('Metrics', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Metrics runtime={runtime} version={version} />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
