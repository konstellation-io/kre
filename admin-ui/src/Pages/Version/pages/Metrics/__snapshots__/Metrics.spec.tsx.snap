// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics matches snapshot 1`] = `
<div
  className="container"
>
  <DashboardHeader
    onChange={[Function]}
    runtimeName="runtimeName"
    setViewAllData={[Function]}
    submit={[Function]}
    value={[Function]}
    versionName="versionName"
    viewAllData={false}
  />
  <Charts
    data={Object {}}
    expanded=""
    toggleExpanded={[Function]}
    viewAllData={false}
  />
</div>
`;
